opcua/common/instantiate.py

```diff
@@ -51,32 +53,49 @@ def _instantiate_node(server, parentid, rdesc, nodeid, bname, recursive=True):
     addnode.TypeDefinition = rdesc.TypeDefinition
 
     node_type = Node(server, rdesc.NodeId)
-    if rdesc.NodeClass in (ua.NodeClass.Object, ua.NodeClass.ObjectType):
-        addnode.NodeClass = ua.NodeClass.Object
-        _read_and_copy_attrs(node_type, ua.ObjectAttributes(), addnode)
-
-    elif rdesc.NodeClass in (ua.NodeClass.Variable, ua.NodeClass.VariableType):
-        addnode.NodeClass = ua.NodeClass.Variable
-        _read_and_copy_attrs(node_type, ua.VariableAttributes(), addnode)
-    elif rdesc.NodeClass in (ua.NodeClass.Method,):
-        addnode.NodeClass = ua.NodeClass.Method
-        _read_and_copy_attrs(node_type, ua.MethodAttributes(), addnode)
+    
+    refs = node_type.get_referenced_nodes(refs=ua.ObjectIds.HasModellingRule)
+    # skip optional elements
+    if not(len(refs) == 1 and refs[0].nodeid == ua.NodeId(ua.ObjectIds.ModellingRule_Optional) ):
+        
+        if rdesc.NodeClass in (ua.NodeClass.Object, ua.NodeClass.ObjectType):
+            addnode.NodeClass = ua.NodeClass.Object
+            _read_and_copy_attrs(node_type, ua.ObjectAttributes(), addnode)
+
+        elif rdesc.NodeClass in (ua.NodeClass.Variable, ua.NodeClass.VariableType):
+            addnode.NodeClass = ua.NodeClass.Variable
+            _read_and_copy_attrs(node_type, ua.VariableAttributes(), addnode)            
+        elif rdesc.NodeClass in (ua.NodeClass.Method,):
+            addnode.NodeClass = ua.NodeClass.Method
+            _read_and_copy_attrs(node_type, ua.MethodAttributes(), addnode)
+        else:
+            print("Instantiate: Node class not supported: ", rdesc.NodeClass)
+            return
+    
+        res = server.add_nodes([addnode])[0]
+        
+        if recursive:
+            parents = ua_utils.get_node_supertypes(node_type, includeitself = True)
+            for parent in parents:
+                descs = parent.get_children_descriptions(includesubtypes=False)
+                for c_rdesc in descs:
+                    #TODO: smells, is there a better way to test if a browse name is already present ?                    
+                    # skip items that already exists, prefer the 'lowest' one in object hierarchy
+                    node = Node(server, res.AddedNodeId)
```

Comment:
create the node outside of loop, you are recreating it many times!. Then add a comment to explain what you are doing like 'checking if child node is already instantiated'. Tok me some time to understand why you did that. Maybe split method and give explicit name.like "is_child_instantiated(node)"